src/main/java/com/yelp/nrtsearch/server/luceneserver/AddDocumentHandler.java

```diff
@@ -92,9 +92,34 @@ public static DocumentsContext getDocumentsContext(
       ((IndexableFieldDef) (IndexState.getMetaField(IndexState.NESTED_PATH)))
           .parseDocumentField(
               documentsContext.getRootDocument(), List.of(IndexState.ROOT), List.of());
+
+      // Include all fields and meta-fields in field names
+      extractFieldNames(documentsContext);
       return documentsContext;
     }
 
+    /** Extract all field names for each document and stores it into a hidden field */
+    private static void extractFieldNames(DocumentsContext documentsContext) {
+      extractFieldNamesForDocument(documentsContext.getRootDocument());
+      documentsContext.getChildDocuments().values().stream()
+          .flatMap(List::stream)
+          .forEach(LuceneDocumentBuilder::extractFieldNamesForDocument);
+    }
+
+    /** Extract all field names in the document and stores it into a hidden field */
+    private static void extractFieldNamesForDocument(Document document) {
+      IndexableFieldDef fieldNamesFieldDef =
+          (IndexableFieldDef) IndexState.getMetaField(IndexState.FIELD_NAMES);
+
+      List<String> fieldNames =
+          document.getFields().stream()
+              .map(IndexableField::name)
+              .distinct()
+              .collect(Collectors.toList());
```

Comment:
maybe filter out the predefined meta fields like _nested_path

Reply:
This is based on https://github.com/elastic/elasticsearch/blob/e144471b3e607cc8b6b3c7a0e7f4706a4bdb9381/server/src/main/java/org/elasticsearch/indices/IndicesModule.java#L184, which explicitly `//_field_names must be added last so that it has a chance to see all the other mappers`. I don't believe adding the meta-fields as terms will significantly affect performance, but we can remove them if needed.

I also tested exists queries on `_id` and `_version` and it works as intended. It could be useful for meta fields we add in the future.